platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/DefaultProjectManager.java

```diff
@@ -439,134 +430,127 @@ public ProjectConfig getProjectConfig(Project project) throws ServerException, P
+                                                                                         ProjectTypeConstraintException,
+                                                                                         ForbiddenException {
+        List<ProjectConfigDto> modules = new ArrayList<>();
 
-                final ValueProviderFactory valueProviderFactory = var.getValueProviderFactory();
+        for (ProjectConfig config : projectConfig.getModules()) {
+            getModulesRecursive(config, modules);
+        }
 
-                // calculate provided values
-                if (valueProviderFactory != null) {
-                    valueProviderFactory.newInstance(project.getBaseFolder()).setValues(var.getName(), attributeValue);
-                }
+        final ProjectConfigDto projectConfigDto = DtoFactory.getInstance().createDto(ProjectConfigDto.class)
+                                                            .withPath(project.getPath())
+                                                            .withName(project.getName())
+                                                            .withModules(modules)
+                                                            .withSource(DtoFactory.getInstance().createDto(SourceStorageDto.class));
 
-                if (attributeValue == null && var.isRequired()) {
-                    throw new ProjectTypeConstraintException("Required attribute value is initialized with null value " + var.getId());
-                }
-
-                // store non-provided values into JSON
-                if (valueProviderFactory == null) {
-                    projectConfig.getAttributes().put(definition.getName(), attributeValue);
-                }
+        ProjectTypes types = new ProjectTypes(project, projectConfig.getType(), projectConfig.getMixins(), this);
+        types.removeTransient();
 
-                checkVariables.put(attributeName, attributeValue);
-            }
-        }
+        projectConfigDto.setType(types.getPrimary().getId());
+        projectConfigDto.setDescription(projectConfig.getDescription());
 
-        for (ProjectType t : types.getAll().values()) {
-            for (Attribute attr : t.getAttributes()) {
-                if (attr.isVariable()) {
-                    // check if required variables initialized
-//                    if(attr.isRequired() && attr.getValue() == null) {
-                    if (!checkVariables.containsKey(attr.getName()) && attr.isRequired()) {
-                        throw new ProjectTypeConstraintException("Required attribute value is initialized with null value " + attr.getId());
-                    }
-                } else {
-                    // add constants
-                    projectConfig.getAttributes().put(attr.getName(), attr.getValue().getList());
-                }
-            }
-        }
+        ArrayList<String> ms = new ArrayList<>();
+        ms.addAll(types.getMixins().keySet());
```

Comment:
You can do it in a single line `new ArrayList<>(types.getMixins().keySet())` which also improves memory usage.